File: server/models/vehicle.model.js
import mongoose from 'mongoose';

const vehicleSchema = new mongoose.Schema({
  vehicleNumber: {
    type: String,
    required: [true, 'Vehicle number is required'],
    uppercase: true,
    trim: true,
    index: true
  },
  vehicleType: {
    type: String,
    required: [true, 'Vehicle type is required'],
    enum: ['car', 'bike', 'bicycle', 'truck', 'bus', 'auto', 'scooter', 'other'],
    lowercase: true
  },
  ownerName: {
    type: String,
    required: [true, 'Owner name is required'],
    trim: true
  },
  ownerRole: {
    type: String,
    required: [true, 'Owner role is required'],
    enum: ['student', 'faculty', 'staff', 'visitor', 'contractor', 'other'],
    lowercase: true
  },
  universityId: {
    type: String,
    trim: true,
    sparse: true
  },
  department: {
    type: String,
    trim: true
  },
  contactNumber: {
    type: String,
    required: [true, 'Contact number is required'],
    trim: true
  },
  entryTime: {
    type: Date,
    required: true,
    default: Date.now
  },
  exitTime: {
    type: Date,
    default: null
  },
  gateName: {
    type: String,
    required: [true, 'Gate name is required'],
    enum: ['GATE 1', 'GATE 2'],
    trim: true
  },
  status: {
    type: String,
    enum: ['inside', 'exited'],
    default: 'inside'
  },
  duration: {
    type: Number, // in minutes
    default: 0
  },
  purpose: {
    type: String,
    trim: true
  },
  verifiedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  notes: {
    type: String,
    trim: true
  },
  ownerPhone: {
    type: String,
    trim: true
  },
  isBlacklisted: {
    type: Boolean,
    default: false
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }
}, {
  timestamps: true
});

vehicleSchema.index({ vehicleNumber: 1, status: 1 });
vehicleSchema.index({ entryTime: -1 });
vehicleSchema.index({ status: 1 });
vehicleSchema.index({ vehicleType: 1 });
vehicleSchema.index({ ownerRole: 1 });
vehicleSchema.index({ gateName: 1 });

vehicleSchema.virtual('formattedDuration').get(function() {
  if (this.duration === 0) return '0 mins';
  
  const hours = Math.floor(this.duration / 60);
  const minutes = this.duration % 60;
  
  if (hours > 0) {
    return `${hours}h ${minutes}m`;
  } else {
    return `${minutes}m`;
  }
});

vehicleSchema.virtual('isInside').get(function() {
  return this.status === 'inside';
});

vehicleSchema.virtual('hasExited').get(function() {
  return this.status === 'exited';
});

vehicleSchema.set('toJSON', { virtuals: true });
vehicleSchema.set('toObject', { virtuals: true });

// Pre-save middleware to calculate duration if exit time is set
vehicleSchema.pre('save', function(next) {
  if (this.exitTime && this.entryTime) {
    this.duration = Math.floor((this.exitTime - this.entryTime) / (1000 * 60));
  }
  next();
});

export default mongoose.model('Vehicle', vehicleSchema);

File: frontend/lib/models/vehicle_model.dart
import 'dart:convert';

class VehicleModel {
  final String id;
  final String vehicleNumber;
  final String vehicleType;
  final String ownerName;
  final String ownerRole;
  final String? universityId;
  final String? department;
  final String contactNumber;
  final DateTime entryTime;
  final DateTime? exitTime;
  final String gateName;
  final String status; 
  final int duration;
  final String? purpose;
  final String? verifiedBy;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  VehicleModel({
    required this.id,
    required this.vehicleNumber,
    required this.vehicleType,
    required this.ownerName,
    required this.ownerRole,
    this.universityId,
    this.department,
    required this.contactNumber,
    required this.entryTime,
    this.exitTime,
    required this.gateName,
    required this.status,
    required this.duration,
    this.purpose,
    this.verifiedBy,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  VehicleModel copyWith({
    String? id,
    String? vehicleNumber,
    String? vehicleType,
    String? ownerName,
    String? ownerRole,
    String? universityId,
    String? department,
    String? contactNumber,
    DateTime? entryTime,
    DateTime? exitTime,
    String? gateName,
    String? status,
    int? duration,
    String? purpose,
    String? verifiedBy,
    String? notes,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return VehicleModel(
      id: id ?? this.id,
      vehicleNumber: vehicleNumber ?? this.vehicleNumber,
      vehicleType: vehicleType ?? this.vehicleType,
      ownerName: ownerName ?? this.ownerName,
      ownerRole: ownerRole ?? this.ownerRole,
      universityId: universityId ?? this.universityId,
      department: department ?? this.department,
      contactNumber: contactNumber ?? this.contactNumber,
      entryTime: entryTime ?? this.entryTime,
      exitTime: exitTime ?? this.exitTime,
      gateName: gateName ?? this.gateName,
      status: status ?? this.status,
      duration: duration ?? this.duration,
      purpose: purpose ?? this.purpose,
      verifiedBy: verifiedBy ?? this.verifiedBy,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      '_id': id,
      'vehicleNumber': vehicleNumber,
      'vehicleType': vehicleType,
      'ownerName': ownerName,
      'ownerRole': ownerRole,
      'universityId': universityId,
      'department': department,
      'contactNumber': contactNumber,
      'entryTime': entryTime.toIso8601String(),
      'exitTime': exitTime?.toIso8601String(),
      'gateName': gateName,
      'status': status,
      'duration': duration,
      'purpose': purpose,
      'verifiedBy': verifiedBy,
      'notes': notes,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  factory VehicleModel.fromMap(Map<String, dynamic> map) {
    return VehicleModel(
      id: map['_id'] ?? map['id'] ?? '',
      vehicleNumber: map['vehicleNumber'] ?? '',
      vehicleType: map['vehicleType'] ?? '',
      ownerName: map['ownerName'] ?? '',
      ownerRole: map['ownerRole'] ?? 'student',
      universityId: map['universityId'],
      department: map['department'],
      contactNumber: map['contactNumber'] ?? '',
      entryTime: DateTime.parse(map['entryTime'] ?? DateTime.now().toIso8601String()),
      exitTime: map['exitTime'] != null ? DateTime.parse(map['exitTime']) : null,
      gateName: map['gateName'] ?? '',
      status: map['status'] ?? 'inside',
      duration: map['duration']?.toInt() ?? 0,
      purpose: map['purpose'],
      verifiedBy: map['verifiedBy'],
      notes: map['notes'],
      createdAt: DateTime.parse(map['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(map['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  String toJson() => json.encode(toMap());

  factory VehicleModel.fromJson(String source) =>
      VehicleModel.fromMap(json.decode(source) as Map<String, dynamic>);

  @override
  String toString() {
    return 'VehicleModel(id: $id, vehicleNumber: $vehicleNumber, vehicleType: $vehicleType, ownerName: $ownerName, status: $status)';
  }

  @override
  bool operator ==(covariant VehicleModel other) {
    if (identical(this, other)) return true;

    return other.id == id &&
        other.vehicleNumber == vehicleNumber &&
        other.vehicleType == vehicleType &&
        other.ownerName == ownerName &&
        other.ownerRole == ownerRole &&
        other.universityId == universityId &&
        other.department == department &&
        other.contactNumber == contactNumber &&
        other.entryTime == entryTime &&
        other.exitTime == exitTime &&
        other.gateName == gateName &&
        other.status == status &&
        other.duration == duration &&
        other.purpose == purpose &&
        other.verifiedBy == verifiedBy &&
        other.notes == notes &&
        other.createdAt == createdAt &&
        other.updatedAt == updatedAt;
  }

  @override
  int get hashCode {
    return id.hashCode ^
        vehicleNumber.hashCode ^
        vehicleType.hashCode ^
        ownerName.hashCode ^
        ownerRole.hashCode ^
        universityId.hashCode ^
        department.hashCode ^
        contactNumber.hashCode ^
        entryTime.hashCode ^
        exitTime.hashCode ^
        gateName.hashCode ^
        status.hashCode ^
        duration.hashCode ^
        purpose.hashCode ^
        verifiedBy.hashCode ^
        notes.hashCode ^
        createdAt.hashCode ^
        updatedAt.hashCode;
  }

  // Utility getters
  String get formattedDuration {
    if (duration == 0) return '0 mins';

    final hours = duration ~/ 60;
    final minutes = duration % 60;
    
    if (hours > 0) {
      return '${hours}h ${minutes}m';
    } else {
      return '${minutes}m';
    }
  }

  bool get isInside => status == 'inside';
  bool get hasExited => status == 'exited';
  
  // Backward compatibility
  bool get isParked => status == 'inside';
  String get plateNumber => vehicleNumber; // For backward compatibility
}
